contracts/airdrop-token-vesting/src/contract.rs

```diff
@@ -0,0 +1,610 @@
+        ExecuteMsg::RewardUsers {
+            rewards,
+            vesting_schedule,
+        } => reward_users(deps, env, info, rewards, vesting_schedule),
+        ExecuteMsg::DeregisterVestingAccount {
+            address,
+            vested_token_recipient,
+            left_vesting_token_recipient,
+        } => deregister_vesting_account(
+            deps,
+            env,
+            info,
+            address,
+            vested_token_recipient,
+            left_vesting_token_recipient,
+        ),
+        ExecuteMsg::Claim { recipient } => claim(deps, env, info, recipient),
+        ExecuteMsg::Withdraw { amount, recipient } => {
+            withdraw(deps, env, info, amount, recipient)
+        }
+    }
+}
+
+/// Allow the contract owner to withdraw the funds of the campaign
+///
+/// Ensures the requested amount is less than or equal to the unallocated amount
+pub fn withdraw(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    amount: Uint128,
+    recipient: String,
+) -> Result<Response, ContractError> {
+    let whitelist = WHITELIST.load(deps.storage)?;
+    let mut unallocated_amount = UNALLOCATED_AMOUNT.load(deps.storage)?;
+    let denom = DENOM.load(deps.storage)?;
+
+    if !whitelist.is_admin(info.sender.clone()) {
+        return Err(StdError::generic_err("Unauthorized").into());
+    }
+
+    let amount_max = min(amount, unallocated_amount);
+    if amount_max.is_zero() {
+        return Err(StdError::generic_err("Nothing to withdraw").into());
+    }
+
+    unallocated_amount = unallocated_amount - amount_max;
+    UNALLOCATED_AMOUNT.save(deps.storage, &unallocated_amount)?;
+
+    // validate recipient address
+    deps.api.addr_validate(&recipient.clone())?;
+
+    Ok(Response::new()
+        .add_messages(vec![build_send_msg(
+            denom,
+            amount_max,
+            recipient.clone(),
+        )?])
+        .add_attribute("action", "withdraw")
+        .add_attribute("recipient", &recipient)
+        .add_attribute("amount", &amount_max.to_string())
+        .add_attribute("unallocated_amount", &unallocated_amount.to_string()))
+}
+
+fn reward_users(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    rewards: Vec<RewardUserRequest>,
+    mut vesting_schedule: VestingSchedule,
+) -> Result<Response, ContractError> {
+    let mut res = vec![];
+
+    let whitelist = WHITELIST.load(deps.storage)?;
+
+    if !(whitelist.is_member(info.sender.clone())
+        || whitelist.is_admin(info.sender.clone()))
+    {
+        return Err(StdError::generic_err("Unauthorized").into());
+    }
+
+    let unallocated_amount = UNALLOCATED_AMOUNT.load(deps.storage)?;
+
+    let total_requested: Uint128 =
+        rewards.iter().map(|req| req.vesting_amount).sum();
+    if total_requested > unallocated_amount {
+        return Err(
+            StdError::generic_err("Insufficient funds for all rewards").into()
+        );
+    }
+    vesting_schedule.validate_time(env.block.time)?;
```

Comment:
It depends. It is a little arbitrary because there is no consistency. Do we want to allow deploying without taking care of if the cliff or the start time is in the past?